README.md
# similarity

### A python package for computing similarity scores between different one-dimensional distributions.

-------

Install by running `python setup.py install` in the similarity directory once unpacked.

-------

#### The theory.

Some example data ships with the package, drawn from simulations presented in Donaldson et al. (2022).

The cumulative distribution functions (CDFs) presented in this work exhibit many similarities. We therefore wish to test the null hypothesis of two CDFs being drawn from the same parent distribution in order to quantify when the distributions are so dissimilar as to {\it not} be drawn from the same distribution. We use a lightly modified Kolmogorov-Smirnov (KS) statistic procedure. As the KS statistic may be difficult to interpret, we detail our procedure and calibrate the values of the KS statistic for our study in this Appendix.

Given two empirical CDFs, $C_1$ and $C_2$, sampled at the same $v$ points, we define the KS statistic as $${\mathcal D} = \max_{1 \leq v \leq N}\left(C_1(v) - C_2(v-1),C_2(v) - C_1(v)\right).$$ As $C_1$ and $C_2$ only have values between 0 and 1, the test statistic has the same range: ${\mathcal D}=0$ corresponds to perfectly matched distributions, and ${\mathcal D}=1$ corresponds to distributions that have exactly no overlap. For traditional KS tests, the ${\mathcal D}$ is compared to the CDF of a normal distribution, and may be straightforwardly translated to a probability given a $\chi^2$ distribution. For our application, we require an empirical calibration of ${\mathcal D}$, which we construct by subsampling the velocity CDFs.

In order to compute the KS statistic, $C_1$ and $C_2$ must be sampled at the same points. To compute matched sets from two unmatched CDFs, $c_1$ and $c_2$ (with number of samples $n_{c_1}$ and $n_{c_2}$ respectively), we proceed as follows. Define a new lower bound $c_{\rm max} = \max\left(\min(c_1),\min(c_2)\right)$, upper bound $c_{\rm min} = \min\left(\max(c_1),\max(c_2)\right)$, and sample resolution $\Delta v=(c_{\rm max}-c_{\rm min})/N$ (for a given choice of $N$), to construct a new set of sample points $V = \left(c_{\rm min},c_{\rm min}+\Delta v, c_{\rm min},c_{\rm min}+2\Delta v,\ldots,c_{\rm max}\right)$. For best results, $n_{c_1}$ and $n_{c_2}$ should have similar numbers of samples, and the choice of $N$ should be similar. However, we find in practice that the choice of sample points varied by a factor of 2 does not strongly influence the results. We empirically estimate the uncertainty from choice of $N$ to be $\Delta {\mathcal D}<0.01$.

To calibrate ${\mathcal D}$, we begin with a 'true' empirical distribution $c_1$ and construct $c_2$ by drawing $n_{c_2}={\rm int}(\alpha n_{c_1})$ where $0<\alpha<1$ random samples from $c_1$. By scanning through $\alpha$ values, we can determine the value of $\alpha$ above which results are converged, and the uncertainty in ${\mathcal D}$. In practice, we are probing two different regimes when calculating ${\mathcal D}$: one where $N\approx500$, and one where $N\approx10000$. We report calibrations for both. For the $N\approx500$ case, we find that ${\mathcal D}$ is converged for $\alpha>0.6$, with $\langle{\mathcal D}\rangle = 0.041\pm0.013$. For the $N\approx10000$ case, we find that ${\mathcal D}$ is also converged for $\alpha>0.6$, with $\langle{\mathcal D}\rangle = 0.010\pm0.003$.

For applications in this work, we theorise that the mean values of two CDFs $c_1$ and $c_2$ may be offset relative to one another. To test the dissimilarity of distributions independent of fixed mean offsets, we compute ${\mathcal D}$ for different offset values $\Delta x$, given two distributions $c_1$ and $c_2+\Delta x$. We test values of $\Delta x$ in a range defined by $\pm2(\langle c_1\rangle-\langle c_2\rangle)$. To calibrate uncertainty, we compute ${\mathcal D}$ for different values of $\Delta x$ using the same CDF for $c_1$ and $c_2$. We find the theoretically-expected relationship ${\mathcal D}\approx |\Delta x|/\langle c_1\rangle$ holds, such that the uncertainty in ${\mathcal D}$ may be approximated as ${\mathcal D}\approx |\langle c_1\rangle-\langle c_2\rangle|/\langle c_1\rangle$. Such a search for a minimum value of ${\mathcal D}$ will maximise the chances of finding that the two CDFs are drawn from the same parent distribution, and give the strongest constraints.

One additional shortcoming of the KS statistic is the insensitivity to subtle variations in the tails of the distribution. However, we can place a theoretical limit on the differences by noting that if one were to truncate the distribution $c_2$ at some maximum $v$, then ${\mathcal D}\approx 1-c_1(v)$. That is, if one defined $c_2$ to be the minimum 90\% of the values of $c_1$, $c_1(v)=0.9$, and we would compute ${\mathcal D}\approx0.1$. We may therefore place meaningful upper limits on dissimilarities in the tails of distributions by computing ${\mathcal D}$.

Given the calibrations, we can make statements about the dissimilarity of two distributions given some simple characteristics. Collecting all of the calibration uncertainties, we find that the uncertainty on the minimum value of ${\mathcal D}$ will be primarily set by the number of samples $n_{c_1}$ and $n_{c_2}$, with other uncertainties strongly subdominant except for at very large values of $n_{c_1}$ and $n_{c_2}$. Defining $n\equiv n_{c_1}\approx n_{c_2}$, for two CDFs $c_1$ and $c_2$, we are unable to distinguish between the two CDFs at the 95\% confidence level if ${\mathcal D}<0.016\sqrt{10000/n}.$


### License
-------

This project is Copyright (c) Michael Petersen and licensed under the terms of the two-clause BSD license. See the ``licenses`` folder for more information.


CHANGES.rst

0.1 (2021 Nov 3)
----------------

- Initial release.


similarity/tests/run_basetests.py
"""
this file tests the basic distributions

"""
import numpy as np
import pkg_resources

# bring in the package itself
import similarity

# identify the testing files
d1 = pkg_resources.resource_filename('similarity','data/LMC.evolved.galactocentric.solar.system1_3eb.dat')
d2 = pkg_resources.resource_filename('similarity','data/LMC.evolved.galactocentric.solar.system1_5eb.dat')

# read in the models, which are 6d positions of dark matter particles in the solar neighbourhood
model1 = np.genfromtxt(d1,delimiter=';',skip_header=1)
model2 = np.genfromtxt(d2,delimiter=';',skip_header=1)

# now bring in the default measurement
from similarity import similarity

vtot1 = np.sqrt(model1[:,3]*model1[:,3] + model1[:,4]*model1[:,4] + model1[:,5]*model1[:,5])
vtot2 = np.sqrt(model2[:,3]*model2[:,3] + model2[:,4]*model2[:,4] + model2[:,5]*model2[:,5])

result = similarity.compare(vtot1,vtot2)
print(result.ks,result.ksDp,result.ksDm,result.p1_bar)


from similarity import empirical

D1 = empirical.Distribution(vtot1)
D2 = empirical.Distribution(vtot1*1.0001)

#print(D1.cdf)

empirical.match_distributions(D1,D2)

# test the calibration classes
from similarity import calibrate

C = calibrate.SingleTest(vtot1)
print('uncertainty from subsampling 1:',C.uncertainty_subsample)

C = calibrate.SingleTest(vtot2)
print('uncertainty from subsampling 2:',C.uncertainty_subsample)

C = calibrate.DoubleTest(vtot1,vtot2,verbose=1)
print('baseline KS',C.base_ks)
print('uncertainty from constructing matched CDFs:',C.uncertainty_spacing)
print('minimum value when shifting CDFs:',C.uncertainty_shifting)

# test the raw KS classes
from similarity import ks

K = ks.KS(D1.hcdf,D2.hcdf)

print(K.ks)
#K._significance()
print(K.alpha)
print('p1bar',K.p1_bar)

print(K._h_lam(0.07,imax=1000))
#print('hlam',K.p1_bar)

# bring in some different analytical distributions
from similarity.analytical import *

velvals = np.linspace(0.,800,1000)
vdisp = 117

maxwellian_shm = Maxwellian(velvals,vdisp)

shm = MultiGaussian(velvals)
shm.trivariate_gaussian_pdf(vdisp*np.sqrt(3),-100.)
shm.trivariate_gaussian_pdf(vdisp*np.sqrt(3),cen1=11.1,cen2=229.+12.24,cen3=7.25)

#mcdf = make_cdf(velvals,maxwellian(velvals,vdisp))
#plt.plot(maxwellian_shm.xvals,maxwellian_shm.pdf)
#plt.plot(shm.xvals,shm.pdf)




from similarity import censored

sample1 = np.array([30,24,11,19,27,11,24,28])
observed1 = np.array([1,0,1,0,1,1,1,0])
sample2 = np.array([3,23,17,8,10,5])
observed2 = np.array([1,1,0,0,1,0])



censored.calculate_twosample(sample1,sample2,observed1,observed2,censoring='right',verbose=1)



similarity/ks.py
"""
the KS test driver

Read the introduction of Hodges (1958), "The significance probability of the Smirnov two-sample test"
P1 is the significance of the 

The Kolmogorov (1933) - Smirnov (1939) test is a rank-order test of a two-sample problem,
where one sample may be assumed to be significantly larger than the other (or not!).

This software is primarily focused on the two-sample test, but can be readily repurposed to perform a one-sample style test.

The null hypothesis is that the two samples are equivalent. Given a binomial coefficient of possible orderings (n1+n2, n2),
each has the same probability, 1/(n1+n2,n).
[compute a binomial coefficient in python as scipy.special.binom(a,b)]

When has a distribution strayed too far from the other such that they cannot be consistent?


Follow Section 4 of Hodges for nearly equal sample sizes.

See also Chapter 26 of DasGupta (2008), Aymptotic Theory of Statistics
Much of the theory here is built on the Empirical Distribution Function (EDF) 


One must map the KS statistic to a P value, using a KS distribution.
The P value is the answer to this question:

If the two samples were randomly sampled from identical populations, what is the probability that the two cumulative frequency distributions would be as far apart as observed? More precisely, what is the chance that the value of the Komogorov-Smirnov D statistic would be as large or larger than observed?

If the P value is small, conclude that the two groups were sampled from populations with different distributions. The populations may differ in median, variability or the shape of the distribution.

pros:
-The test is nonparametric. It does not assume that data are sampled from Gaussian distributions (or any other defined distributions).
-The results will not change if you transform all the values to logarithms or reciprocals or any transformation. The KS test report the maximum difference between the two cumulative distributions, and calculates a P value from that and the sample sizes. A transformation will stretch (even rearrange if you pick a strange transformation) the X axis of the frequency distribution, but cannot change the maximum distance between two frequency distributions.

cons:
The primary shortcoming in this approach is that we are not particularly sensitive to changes in the tails of the distributions.



"""

import numpy as np

class KS():

    def __init__(self,y1,y2):

        n1,n2 = len(y1),len(y2)
    
        if n1 != n2:
            print('mismatched sizes!');return

        self.y1 = y1
        self.y2 = y2

        self.n1 = n1
        self.n2 = n2

        # compute all statistics
        self._compute_ks()     # two-sided test
        self._compute_ks_one() # one-sided test
        self._p1_bar()         # p1 bar
        self._compute_alpha()  # significance level

    def _compute_ks(self):
        """compute the formal two-sided KS test
    
        see https://www.itl.nist.gov/div898/handbook/eda/section3/eda35g.htm

        in Hodges, this is 'D'.
        The null hypothesis is that the distributions are identical; nonzero values indicate they are not identical (with varying certainty).

        The mathematical operation is called the 'supremum'.
        """

        ks = 0.
        for i in range(1,self.n1):
            testval = np.nanmax([self.y1[i]-self.y2[i-1],self.y2[i]-self.y1[i]])
            if testval > ks:
                ks = testval
            
        self.ks = ks

    def _compute_ks_one(self):
        """compute the formal one-sided KS test
    
        in Hodges, this is 'D+' and 'D-'.


        These are the one-sided tests. They examine the null hypotheses that
        D+: y1(x) >= y2(x) for all x; the alternative is that y1(x) < y2(x) for at least one x.
        D-: y2(x) >= y1(x) for all x; the alternative is that y2(x) < y1(x) for at least one x.
        """

        ks = 0.
        for i in range(1,self.n1):
            testval = np.nanmax([self.y1[i]-self.y2[i-1]])
            if testval > ks:
                ks = testval
            
        self.ksDp = ks

        ks = 0.
        for i in range(1,self.n1):
            testval = np.nanmax([self.y2[i]-self.y1[i]])
            if testval > ks:
                ks = testval
            
        self.ksDm = ks

    def _p1_bar(self):
        """
        Hodges eq. 5.1, an estimate for fluctuations when the samples are nearly equal (or equal!)

        The remainder here is of order 1/n

        Implement Hodges 5.3 as well.

        This is the significance value we should be targeting to clear in order to disprove the null hypothesis.
        This is a combinatorial approach: we also have empirical calibrations.
        """

        z = self.n1/self.n2

        # Hodges 5.1
        #self.p1_bar = np.exp(-2*z*z - np.sqrt(2)*z/np.sqrt(self.n2))

        # Hodges 5.1
        self.p1_bar = np.exp(-2*z*z - (2*z/3.)*((self.n1+2.*self.n2)/np.sqrt(self.n1*self.n2*(self.n1+self.n2))))

    def _compute_alpha(self):
        """find the significance value, alpha, based on the ks two-sample test

        the null hypothesis is rejected at percentage alpha if
        ks > c(alpha) sqrt((n1+n2)/(n1*n2)) -> c(alpha) sqrt(2/n1)

        [assuming n1=n2]

        where
        c(alpha) = sqrt( -ln(alpha/2)/2 )

        
        """

        self.alpha = 2.*np.exp(-2.*(self.ks/np.sqrt(2./self.n1))**2.)
        

    def _smirnovs_variables(self):
        """
        following Hodges Section 2, define some variables

        eq. 2.6 of Hodges

        this reduces to _h_lam if done wisely! (I think)
        and the sample size factors are taken into account
        """

        bigZ    = np.sqrt((n1*n2)/(n1+n2))*self.ks
        zplus   = np.sqrt((n1*n2)/(n1+n2))*self.ksD

        self.kz = 1. - 2.*(np.exp(-2.*(1.*bigZ)**2)-
                         np.exp(-2.*(2.*bigZ)**2)+
                         np.exp(-2.*(3.*bigZ)**2)-
                         np.exp(-2.*(4.*bigZ)**2)+
                         np.exp(-2.*(5.*bigZ)**2)-
                         np.exp(-2.*(6.*bigZ)**2))

    def _h_lam(self,lam,imax=1000):
        """return H(t), the cdf of the KS distribution, for significance value to test, lamda.

        in the limit where n1,n2 -> \infty,

        P_{H_0}\left(\sqrt{n1*n2/(n1+n2)}D_{n1,n2}\le \lambda = P( supremum |B(t)| \le \lambda)

        B(t) is a Brownian bridge, which is a whole thing in and of itself!

        In the large sample limit, we can invert this to find out how likely we are to reject the null hypothesis
        e.g. if we want a p-value of 0.05, then the critical value of D_n = 1.35/sqrt(n)

        """

        ivals = np.arange(1,imax,1)
        sumval = np.nansum( -1.**(ivals-1)*np.exp(-2*ivals*ivals*lam*lam))
        return (1.-2.*sumval)/np.sqrt(self.n1)

    



similarity/calibrate.py
"""
calibration of the KS test

for a single distribution
1. subsample distribution against itself

for two distributions:
1. slide distributions against each other for minimum
2. test different grid sizes for interpolation error

"""

import numpy as np

from . import empirical
from . import ks

class SingleTest():
    def __init__(self,Set1,retall=False):

        self.y1 = Set1
        self.n1 = Set1.size

        self.retall = retall

        # create a distribution
        self.D1 = empirical.Distribution(self.y1)

        self._subsample()

    def _subsample(self,nfracs=10,floor_threshold=0.6):
        """a test of discreteness noise in the numbers

        the resulting distributions will decrease monotonically until they hit a floor.
        the scatter about the floor is an estimate in the uncertainty owing to the discreteness in the distribution.

        this process is a bit slow: can we speed it up somehow through code organisation?

        the results are also weirdly unstable? may need more spacing, or to reconsider the combinatorial aspects.
        """

        sample_percentiles = np.linspace(0.1,1.,nfracs)
        ks_values          = np.zeros(nfracs)

        for indx,nfrac in enumerate(sample_percentiles):
            subsample = self.y1[np.random.randint(0,self.n1,int(np.floor(nfrac*self.n1)))]

            # sort...
            #subsample = subsample[subsample.argsort()]

            # create a distribution
            D2 = empirical.Distribution(subsample)

            # compute a homogeneous grid between the two, based on the original spacing.
            empirical.match_distributions(self.D1,D2,spacing=self.D1.dcx)

            # compute the ks values for the two distributions
            K = ks.KS(self.D1.hcdf,D2.hcdf)
            ks_values[indx] = K.ks

        # define the threshold for the noise floor, conservatively 60% of the overall sample size
        criterion = (sample_percentiles>=floor_threshold)

        floor_value   = np.nanmedian(ks_values[criterion])
        floor_scatter =    np.nanstd(ks_values[criterion])

        self.uncertainty_subsample = floor_scatter

        if self.retall:
            self.percentiles = sample_percentiles
            self.ks_values   = ks_values

        


class DoubleTest():
    def __init__(self,Set1,Set2,verbose=0):

        self.y1 = Set1
        self.y2 = Set2
        self.verbose = verbose

        # create the distributions
        self.D1 = empirical.Distribution(self.y1)
        self.D2 = empirical.Distribution(self.y2)

        # get our baseline
        empirical.match_distributions(self.D1,self.D2)

        # compute the KS statistic(s)
        K = ks.KS(self.D1.hcdf,self.D2.hcdf)
        self.base_ks = K.ks

        self._test_spacing()
        self._test_shift()

    def _test_spacing(self):

        base_spacing = np.nanmax([self.D1.dcx,self.D2.dcx])

        spacing_range = np.array([0.1,0.125,0.25,0.5,1,2,4,8])
        ks_values     = np.zeros(spacing_range.size)
        
        for indx,multiplier in enumerate(spacing_range):
            empirical.match_distributions(self.D1,self.D2,spacing=base_spacing*multiplier)

            K = ks.KS(self.D1.hcdf,self.D2.hcdf)
            ks_values[indx] = K.ks

        self.uncertainty_spacing = np.nanstd(ks_values)

        # only for some serious debugging
        if self.verbose>1:
            print('calibrate.DoubleTest._test_spacing: max uncertainty at spacing={0:4.3f}, KS={0:4.3f}'.format(spacing_range[np.nanargmax(ks_values)],np.nanmax(ks_values)))
        

    def _test_shift(self):
        """
        find the minimum KS value when offsetting the distributions relative to one another.
        
        DOES NOT APPLY TO ALL SITUATIONS!

        """

        # establish a maximum shift: very generous here!
        base_shift = np.abs(np.nanmedian(self.y1) - np.nanmedian(self.y2))
        
        max_shift = np.nanmax([np.abs(np.nanpercentile(self.y1,75) - np.nanpercentile(self.y2,25)),
                               np.abs(np.nanpercentile(self.y2,75) - np.nanpercentile(self.y1,25))])

        # how to test for multiple minima?

        # first step, just quote global minumum

        shift_vals = np.linspace(-max_shift,max_shift,30)
        ks_values  = np.zeros(shift_vals.size)
        
        for indx,shift in enumerate(shift_vals):
            # shift all values of D2 relative to D1
            D2s = empirical.Distribution(self.y2 - shift)
            empirical.match_distributions(self.D1,D2s)

            K = ks.KS(self.D1.hcdf,D2s.hcdf)
            ks_values[indx] = K.ks

        self.uncertainty_shifting = np.nanmin(ks_values)




similarity/empirical.py
"""
class structure for handling different distributions from sets of (x,y) points, with optional weights

"""

import numpy as np

from scipy.interpolate import interp1d



class Distribution():
    """
    class to handle empirical distributions, making pdfs and cdfs from
    individual observations
    """

    def __init__(self,xvalues,weights=1,xbins=-1):
        """
        create empirical distributions from observations

        inputs
        ---------
        xvalues : (array of floats) the x-positions of the observations
        yvalues : (array of floats) the y-values of the observations
        weights : (array of floats, optional) the weights of the observations
        xbins   : (array of floats, optional) the points to bin to

        """

        self.xvalues = xvalues

        if isinstance(weights,int):
            self.weights = np.ones(xvalues.size)
        else:
            # check if the length matches xvalues?
            self.weights = weights

        # if xbins is > 0, use as the binning range. otherwise, make
        # an adaptive binning range
        if isinstance(xbins,int):
            if xbins<0:
                self.xbins = np.linspace(np.nanmin(xvalues),np.nanmax(xvalues),int(np.sqrt(xvalues.size)))
            else:
                self.xbins = np.linspace(np.nanmin(xvalues),np.nanmax(xvalues),xbins)
        else:
            self.xbins = xbins

        self.dx = self.xbins[1]-self.xbins[0]

        # separate binning for the cumulative distributions
        self.xcbins = np.linspace(np.nanmin(xvalues),np.nanmax(xvalues),xvalues.size)
        self.dcx = self.xcbins[1]-self.xcbins[0]

        # make the cdf and pdf
        self._weighted_pdf()
        self._weighted_cdf()

        # set up for homogenisation with another distribution
        self.hxbins = None
        self.hpdf   = None
        self.hcdf   = None


    def _weighted_pdf(self):
        """
        make a very quick 1d pdf, with weights
        """
        vals = np.zeros_like(self.xbins)
        
        for b in range(0,self.xbins.size):
            if b==self.xbins.size-1:
                w = np.where( (self.xvalues>self.xbins[b]) & (self.xvalues<=self.xbins[b]+self.dx))[0]
            else:
                w = np.where( (self.xvalues>self.xbins[b]) & (self.xvalues<=self.xbins[b+1]))[0]
            vals[b] = np.nansum(self.weights[w])

        # normalise the area to one
        norm = np.nansum(self.dx*vals)

        self.pdf = vals/norm


    def _weighted_cdf(self):
        """
        make a very quick 1d cdf, with weights
        """
        vals = np.zeros_like(self.xcbins)
        for b in range(0,self.xcbins.size):
            if b==self.xcbins.size-1:
                w = np.where((self.xvalues<=self.xcbins[b]+self.dcx))[0]
            else:
                w = np.where((self.xvalues<=self.xcbins[b+1]))[0]
            vals[b] = np.nansum(self.weights[w])
            
        # normalise the area to one
        norm = np.nanmax(vals)
        
        self.cdf = vals/norm


    def _un_weighted_cdf(self):
        """make an unweighted cdf, with full resolution"""
        self.vsort = self.yvalues[self.yvalues.argsort()]
        self.wsort = np.cumsum(self.weights[self.yvalues.argsort()])/np.nansum(self.weights)    


    def _interpolate_to_grid(self,grid):
        """put distributions on a specified grid"""

        self.hxbins = grid
        self.hpdf   = interp1d(self.xbins, self.pdf, kind='linear')(self.hxbins)
        self.hcdf   = interp1d(self.xcbins, self.cdf, kind='linear')(self.hxbins)



def match_distributions(Distribution1,Distribution2,spacing=-1.,verbose=0):
    """create a uniform grid
    
    spacing sets the km/s sampling. 1 km/s is a good default.
    """
    
    if spacing < 0:
        spacing = np.nanmax([Distribution1.dcx,Distribution2.dcx])
        if verbose: print('similarity.empirical.match_distributions: spacing=',spacing)
    
    xnew = np.arange(np.nanmax([np.nanmin(Distribution1.xbins),np.nanmin(Distribution2.xbins)]),
                     np.nanmin([np.nanmax(Distribution1.xbins),np.nanmax(Distribution2.xbins)]),
                     spacing)
    
    Distribution1._interpolate_to_grid(xnew)
    Distribution2._interpolate_to_grid(xnew)
    





similarity/analytical.py
"""
class structure for some basic analytic distributions

Each distribution comes with a pdf and cdf, sampled at points as specified in the call.
Additional parameters are required as specified for each distribution.

supported distributions:
-Chi^2
-Gaussian
-Lognormal
-Maxwellian
-Multivariate Gaussian -> primarily 3d
-Normal (alias for Gaussian)



"""

import numpy as np

from scipy.special import erf,gamma,gammainc



class Chi2():
    def __init__(self,xvals,k=1):

        self.xvals = xvals
        self.dx    = xvals[1]-xvals[0]
        self.k     = k

        self._chi2_pdf()
        self._chi2_cdf()

    def _chi2_pdf(self):
        """chi2 distribution pdf"""

        norm = 2.**(self.k/2.)*gamma(self.k/2.)
        
        tpdf = (self.xvals**(self.k/2. - 1) * np.exp(-self.xvals/2.))/norm

        self.pdf = tpdf#/np.nansum(self.dx*tpdf)

    def _chi2_cdf(self):
        """chi2 distribution cdf"""

        norm = gamma(self.k/2.)
        
        # the scipy definition includes the norm: so don't include here!
        tcdf = gammainc(self.k/2.,self.xvals/2.)#/norm

        self.cdf = tcdf

        

class Gaussian():
    """the OG. The distribution of a random variable with unknown distribution.
    
    owes to the central limit theorem: under some conditions, the average of many sample of a random variable,
    with finite mean and variance (which is itself a random variable), will converge to a normal distribution
    as the number of samples increases.

    """
    def __init__(self,xvals,mu=0.,sigma=1.):

        self.xvals = xvals
        self.mu    = mu
        self.sigma = sigma

        self._gaussian_pdf()
        self._gaussian_cdf()

    def _gaussian_pdf(self):
        """gaussian distribution pdf"""
        self.pdf = (1./self.sigma/np.sqrt(2.*np.pi)) * np.exp(-0.5*((self.xvals-self.mu)/self.sigma)**2.)

    def _gaussian_cdf(self):
        """gaussian distribution cdf"""
        self.cdf = (1./2.) * (1.+erf((self.xvals-self.mu)/(np.sqrt(2)*self.sigma)))

        
class Lognormal():
    def __init__(self,xvals,mu=0.,sigma=1.):

        self.xvals = xvals
        self.mu    = mu
        self.sigma = sigma

        self._lognormal_pdf()

    def _lognormal_pdf(self):
        """gaussian distribution pdf"""
        self.pdf (1./self.sigma/self.xvals/np.sqrt(2.*np.pi)) * np.exp(-0.5*((np.log(self.xvals)-self.mu)/self.sigma)**2.)

                

class Maxwellian():
    def __init__(self,xvals,sigma=1.):

        self.xvals = xvals
        self.sigma = sigma

        self._maxwellian_pdf()
        self._maxwellian_cdf()
        
    def _maxwellian_pdf(self):
        """maxwellian distribution pdf"""
        self.pdf = np.sqrt(2./np.pi)*self.xvals*self.xvals*np.exp(-(self.xvals*self.xvals/2./self.sigma/self.sigma))/self.sigma/self.sigma/self.sigma

    def _maxwellian_cdf(self):
        """maxwellian distribution cdf"""
        self.cdf = erf(self.xvals/np.sqrt(2.)/self.sigma) - np.sqrt(2./np.pi)*self.xvals*np.exp(-(self.xvals*self.xvals/2./self.sigma/self.sigma))/self.sigma


class MultiGaussian():

    def __init__(self,xvals):

        self.xvals = xvals
        
    def trivariate_gaussian_pdf(self,sigma,cen1=0.,cen2=0.,cen3=0.):
        """trivariate gaussian distribution pdf, with uniform
        widths, but optional different centroids"""
        sigma1 = sigma
        sigma2 = sigma
        sigma3 = sigma
        norm   = 4.*np.pi/(sigma1*sigma2*sigma3)/3.
        
        pdf = norm*(self.xvals**2.)*np.exp(
            -((self.xvals-cen1)**2./(2.*sigma1**2.))
            -((self.xvals-cen2)**2./(2.*sigma2**2.))
            -((self.xvals-cen3)**2./(2.*sigma3**2.)))
        
        self.pdf = pdf/np.nansum(pdf*(self.xvals[1]-self.xvals[0]))

        self._make_cdf()
    
    def _make_cdf(self):
        """make a numerical cdf from a pdf"""
        self.cdf = np.cumsum(self.pdf*(self.xvals[1]-self.xvals[0]))



class Normal():
    """an alias for Gaussian"""
    def __init__(self,xvals,mu=0.,sigma=1.):
       G = Gaussian(xvals,mu,sigma)
       self.xvals = G.xvals
       self.mu    = G.mu
       self.sigma = G.sigma
       self.pdf   = G.pdf
       self.cdf   = G.cdf

       


similarity/similarity.py
"""
the similarity score wrapper

more reading:
- https://www.itl.nist.gov/div898/handbook/eda/section3/eda35g.htm
- https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.kstest.html#scipy.stats.kstest
- https://en.wikipedia.org/wiki/Sign_test
- https://en.wikipedia.org/wiki/Logistic_regression

"""

import numpy as np

from . import empirical
from . import ks

class compare():
    """
    """
    def __init__(self,Set1,Set2):
        """
        given two different sets of values, compute the similarity score and
        estimates of associated uncertainties

        this wrapper will use all defaults under the hood -- you may want to access the classes directly for your application.

        inputs
        --------
        Set1
        Set2

        """

        self.similarity = 0.
        self.significance = 0.

        D1 = empirical.Distribution(Set1)
        D2 = empirical.Distribution(Set2)

        # homogenise to share grids
        empirical.match_distributions(D1,D2)

        # compute the KS statistic(s)
        K = ks.KS(D1.hcdf,D2.hcdf)

        # transfer to the compare namespace
        self.ks = K.ks
        self.ksDp = K.ksDp
        self.ksDm = K.ksDm
        self.p1_bar = K.p1_bar
        


similarity/censored.py
"""
Kaplan & Meier (1958) product estimator, as described in Feigelson & Nelson (1985).

This formalism was used in Petersen et al. (2019)




# a very simple example
sample1 = np.array([30,24,11,19,27,11,24,28])
observed1 = np.array([1,0,1,0,1,1,1,0])
sample2 = np.array([3,23,17,8,10,5])
observed2 = np.array([1,1,0,0,1,0])



calculate_twosample(sample1,sample2,observed1,observed2,censoring='right',verbose=1)

"""


import numpy as np

from scipy.stats import chi2


    
def two_sample_prep(sample1,sample2,observed1,observed2,censoring='left'):
    '''
    
    inputs
    ------------
    sample1    : numpy array, observed values or upper limits for first sample
    sample2    : numpy array, observed values or upper limits for first sample
    observed1  : numpy array, same size as sample1. each entry is 1 for detections, 0 for upper limits
    observed2  : numpy array, same size as sample2. each entry is 1 for detections, 0 for upper limits
    censoring  : 'left' or 'right', indicating upper or lower limits
    
    '''
    
    if censoring=='left':
        # find the offsetting value (maximum observed value)
        offset_val = np.max([np.max(sample1[observed1==1]),np.max(sample2[observed2==1])])

        sample1 = offset_val-sample1
        sample2 = offset_val-sample2
    else:
        #
        # right censored is the base assumption. 
        #   most survival analysis is written for right censoring, e.g. death during trial
        #
        print('two_sample_prep: Assuming right-censoring (lower limits).')

    # parition data
    true1 = sample1[observed1==1]
    censored1 = sample1[observed1==0]

    true2 = sample2[observed2==1]
    censored2 = sample2[observed2==0]

    # make the joint sample and rankings
    jsample = np.unique(np.concatenate([sample1[observed1==1],sample2[observed2==1]]))
    jranking = jsample.argsort()
    #print(jsample[jranking])

    # calculate basic statistics
    N1 = sample1.size
    N2 = sample2.size
    rank1 = sample1.argsort()
    rank2 = sample2.argsort()
    n = N1 + N2
    r = np.max([N1,N2]) - 1
    #print(N1,N2,n,r)
    
    return jsample,sample1,true1,censored1,sample2,true2,censored2




def calculate_twosample(sample1,sample2,observed1,observed2,censoring='right',test='peto',verbose=0):
    '''
    
    inputs
    ------------
    sample1    : numpy array, observed values or upper limits for first sample
    sample2    : numpy array, observed values or upper limits for first sample
    observed1  : numpy array, same size as sample1. each entry is 1 for detections, 0 for upper limits
    observed2  : numpy array, same size as sample2. each entry is 1 for detections, 0 for upper limits
    censoring  : 'left' or 'right', indicating upper or lower limits
    test       : 'peto','gehan','logrank', indicating which statistical test to undertake
    verbose    : verbosity flag.
    
    
    
    '''
    
    
    jsample,sample1,true1,censored1,sample2,true2,censored2 = two_sample_prep(sample1,sample2,observed1,observed2,censoring=censoring)


    logrank_L_sum = 0
    logrank_sig_sum = 0
    gehan_L_sum = 0
    gehan_sig_sum = 0
    peto_L_sum = 0
    peto_sig_sum = 0

    peto_w = np.zeros(jsample.size)

    for indx,j in enumerate(jsample):

        n1j = len( np.where(sample1 >= j)[0]) 
        n2j = len( np.where(sample2 >= j)[0])
        nj  = n1j + n2j
        peto_w[indx] = nj/(nj+1)


        #if indx+1 < len(jsample):
        #    m1j = len( np.where( (censored1 >= j) & (censored1 < jsample[indx+1]))[0])
        #    m2j = len( np.where( (censored2 >= j) & (censored2 < jsample[indx+1]))[0])
        #    mj  = m1j + m2j

        #else:
        #    m1j = m2j = mj = 0.

        d1j = len( np.where(true1 == j)[0])
        d2j = len( np.where(true2 == j)[0])
        dj  = d1j + d2j

        # calculate the peto-prentice weight
        wj = np.prod(peto_w[0:indx])

        if indx+1 < len(jsample):
            # now use wj * (d1j - dj*n1j/nj)
            #   for gehan, wj= nj
            gehan_L   = nj * (d1j - (dj*n1j/nj))
            gehan_sig = nj*nj*dj * ((n1j*n2j)/(nj*nj)) * ((nj-dj)/(nj-1))


            logrank_L   = (d1j - (dj*n1j/nj))
            logrank_sig = dj * ((n1j*n2j)/(nj*nj)) * ((nj-dj)/(nj-1))

            peto_L = wj * (d1j - (dj*n1j/nj))
            peto_sig = wj*wj*dj * ((n1j*n2j)/(nj*nj)) * ((nj-dj)/(nj-1))

        else:
            gehan_L = gejan_sig = logrank_L = logrank_sig = 0.

        gehan_L_sum += gehan_L
        gehan_sig_sum += gehan_sig

        logrank_L_sum += logrank_L
        logrank_sig_sum += logrank_sig

        peto_L_sum += peto_L
        peto_sig_sum += peto_sig

        #if verbose > 1:
        #    print(j,n1j,n2j,nj,m1j,m2j,mj,d1j,d2j,dj,logrank_L,logrank_sig)

    if verbose > 0:
        print('Logrank:       L={0:6.2f} s={1:6.2f} p={2:5.4f}'.format(logrank_L_sum,logrank_sig_sum,chi2.sf((logrank_L_sum*logrank_L_sum)/logrank_sig_sum,1)))
        print('Peto-Prentice: L={0:6.2f} s={1:6.2f} p={2:5.4f}'.format(peto_L_sum,peto_sig_sum,chi2.sf((peto_L_sum*peto_L_sum)/peto_sig_sum,1)))
        print('Gehan:         L={0:6.2f} s={1:6.2f} p={2:5.4f}'.format(gehan_L_sum,gehan_sig_sum,chi2.sf((gehan_L_sum*gehan_L_sum)/gehan_sig_sum,1)))
        
    if test == 'gehan':
        return chi2.sf((gehan_L_sum*gehan_L_sum)/gehan_sig_sum,1)
    elif test == 'logrank':
        return chi2.sf((logrank_L_sum*logrank_L_sum)/logrank_sig_sum,1)
    else:
        return chi2.sf((peto_L_sum*peto_L_sum)/peto_sig_sum,1)

    
